airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java

```diff
@@ -227,26 +232,229 @@ private Builder<String, Object> generateMetadata(final StandardSync standardSync
+  private static AirbyteCatalog toPatchedAirbyteCatalog(final AirbyteCatalog existing, final AirbyteCatalog patch) {
+
+    final Map<StreamDescriptor, AirbyteStreamAndConfiguration> existingDescriptorToStreamAndConfig = getDescriptorToStreamAndConfigMap(existing);
+    final Map<StreamDescriptor, AirbyteStreamAndConfiguration> patchDescriptorToStreamAndConfig = getDescriptorToStreamAndConfigMap(patch);
+
+    final Map<StreamDescriptor, AirbyteStreamAndConfiguration> merged =
+        Stream.of(existingDescriptorToStreamAndConfig, patchDescriptorToStreamAndConfig)
+            .flatMap(map -> map.entrySet().stream())
+            .collect(Collectors.toMap(
+                Map.Entry::getKey,
+                Map.Entry::getValue,
+                // if a stream is present in both the existing map and the patch map, keep the patched value
+                (existingVal, patchedVal) -> patchedVal));
+
+    final List<AirbyteStreamAndConfiguration> sortedMergedStreams = sortStreamsForPatchedCatalog(merged.values().stream().toList(), existing, patch);
+
+    return new AirbyteCatalog().streams(sortedMergedStreams);
+  }
+
+  private static Map<StreamDescriptor, AirbyteStreamAndConfiguration> getDescriptorToStreamAndConfigMap(final AirbyteCatalog catalog) {
+    return catalog.getStreams()
+        .stream()
+        .collect(Collectors.toMap(
+            s -> getStreamDescriptorForStream(s.getStream()),
+            s -> s));
+  }
+
+  private static StreamDescriptor getStreamDescriptorForStream(final AirbyteStream stream) {
+    return new StreamDescriptor().name(stream.getName()).namespace(stream.getNamespace());
+  }
+
+  private static Map<StreamDescriptor, Integer> getStreamPositionMap(final AirbyteCatalog catalog) {
+    final Map<StreamDescriptor, Integer> positions = new HashMap<>();
+    for (int i = 0; i < catalog.getStreams().size(); i++) {
+      positions.put(getStreamDescriptorForStream(catalog.getStreams().get(i).getStream()), i);
+    }
+    return positions;
+  }
+
+  /**
+   * Given an unsorted list of streams, the original catalog, and the patch catalog, returns a sorted
+   * list of streams where existing stream positions are preserved, and new streams from the patch are
+   * added in the order they were listed in the patch.
+   */
+  private static List<AirbyteStreamAndConfiguration> sortStreamsForPatchedCatalog(final List<AirbyteStreamAndConfiguration> mergedStreams,
+                                                                                  final AirbyteCatalog existing,
+                                                                                  final AirbyteCatalog patch) {
 
-    configRepository.writeStandardSync(newConnection);
+    final Map<StreamDescriptor, Integer> originalStreamDescriptorPositions = getStreamPositionMap(existing);
+    final Map<StreamDescriptor, Integer> patchStreamDescriptorPositions = getStreamPositionMap(patch);
+
+    return mergedStreams.stream().sorted((stream1, stream2) -> {
```

Comment:
just a thought: if we pull this out into it's own function, it'll be easier to test the sorting logic. Not sure how relevant this is given the existing connection handler tests.

Reply:
Wrote a test for this whole method rather than just the sort, I think it better demonstrates the intent of the behavior while also covering the sort behavior itself

Reply:
Sounds good!